### DISCUSSION TITLE
Carddav with Radicale

### DISCUSSION THREAD

**Discussion by be-root** [blind-coder/rcmcarddav#338] [Q&A]
Hello!
I use carddav (latest) with radicale (latest).
So far this also worked well. Now I wanted to change a contact, but did not work. Note was I think that in the address book path should not be specified localhost or something like that. I wanted to change this but now carddav no longer finds an address book.

so far:
https://localhost:5232/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/

tried:
https://127.0.0.1:5232/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/
or:
https://domain.com:5232/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/

With DAVx5 it works fine. Therefore it can't be the address books of radicale.
Where is the error?
I am grateful for any hint!

**Comment by mstilkerich** [comment]
Hi,

  I don't fully understand the issue. It is currently not possible to change the URL of an addressbook, you have to delete it and create it as a new one with the new URL. If you only want to change the URL because you cannot create contacts anymore though, this should not be needed. localhost should resolve fine on the webserver, assuming of course that radicale runs on the webserver :-)

I just tried on my test environment to create a new contact with radicale and rcmcarddav 4.1.0, which works fine for me.

Can you provide a logfile (logs/carddav.log, logs/carddav_http.log) that should show the error? Please set the loglevel to DEBUG for both carddav and http logging and then try to create a contact.

The debug level is configured in `plugins/carddav/config.inc.php`, adapt
```php
$prefs['_GLOBAL']['loglevel'] = \Psr\Log\LogLevel::DEBUG;                                                              
$prefs['_GLOBAL']['loglevel_http'] = \Psr\Log\LogLevel::DEBUG;
```

**Comment by be-root** [comment]
Hi, thanks for your quick reply!
It seems to be a curl problem:

https://localhost:5232/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/

curl: (60) SSL: no alternative certificate subject name matches target host name 'localhost'.

http://localhost:5232/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/

curl: (52) Empty reply from server
(I don't know why...)

https://domain.com:5232/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/

curl: (7) Failed to connect to ... port 443: The waiting time for the connection has expired

What could this be?

**Comment by mstilkerich** [comment]
Hello,

  I suppose you enabled SSL on the radicale server and now roundcube does not work anymore, at least that is plausible for your first two settings. The last one is weird because it attempted to connect to port 443 although the URL you showed explicitly specifies port 5232. I tried this on my end, however, and it works fine.

So:
- If you are using SSL, you cannot use localhost because you normally won't have a certificate issued to localhost. You need to have the server name in the URL that is also used in the certificate.
- The empty reply from server is plausible if you try to connect to an HTTPS server by plain HTTP. I could reproduce this on my end.
- Assuming you have the Radicale server with `ssl=true` running at carddav.domain.com on port 5232, add the new addressbook by using `https://carddav.domain.com:5232` as the URL. It will discover the correct path automatically.

As I wrote, I tried this on my end with rcmcarddav 4.1.0 and it works fine. If it does not for you, please provide a carddav_http.log that shows what connection attempts rcmcarddav actually makes and where things go wrong. Also, your radicale config would be helpful to ensure the used port and that ssl is enabled.

**Reply by be-root** [reply]
Thank you for your support so far!
RC 1.5-beta, CardDAV 4.1, PHP8.0

LOG:
--------
cURL error 7: Failed to connect to 192.168.2.2 port 5235: Connection refused (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for http://192.168.2.2:5235/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/
[15-Mar-2021 18:19:26 +0100]: <ah089cpd> [5 ERR] "PROPFIND /user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/ HTTP/1.1" NULL
>>>>>>>>
PROPFIND /user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/ HTTP/1.1
Content-Length: 215
User-Agent: GuzzleHttp/7
Host: domain.com
Depth: 0
Content-Type: application/xml; charset=UTF-8
Prefer: return=minimal

<?xml version="1.0"?>
<DAV:propfind xmlns:DAV="DAV:" xmlns:CARDDAV="urn:ietf:params:xml:ns:carddav" xmlns:CS="http://calendarserver.org/ns/">
 <DAV:prop>
  <DAV:current-user-principal/>
 </DAV:prop>
</DAV:propfind>

<<<<<<<<

--------
cURL error 7: Failed to connect to domain.com port 443: Connection timed out (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://domain.com:5232/user/fg353tg4-hj65-rg5g-h5f4-3425fewfwef/

Could it be the PHP version?
Thanks a lot!

**Reply by mstilkerich** [reply]
Hello,

  your logs make no sense to me: The first entry uses port 5235, the second has port 443 inside the curl error message and 5232 in the URL it shows afterwards. The first entry has an IP address, the second has a domain name. What is correct now? Are these logs really produced for a single addressbook create operation?

I don't think it's the PHP version, because I have PHP 8 running on the github CI systems and haven't had  any failing tests for it.

So to progress here, please provide __all__ of the following information:
  - Which URL are you using successfully with DavX5? (the same URL should of course also work with rcmcarddav)
  - Just to be sure, can you please show your radicale config file? On my development VM it looks like this:
  ```
[server]
hosts = 127.0.0.1:5232, [::1]:5232
ssl = true
key = /etc/ssl/private/forge.mike2k.de/forge.mike2k.de.key
certificate = /etc/ssl/private/forge.mike2k.de/fullchain.cer

[auth]
type = htpasswd
htpasswd_filename = /etc/radicale/users
htpasswd_encryption = md5

[storage]
filesystem_folder = /srv/radicale/collections
  ```
  - What URL did you enter in rcmcarddav to create the addressbook?
  - The full carddav.log and carddav_http.log files for adding the addressbook. (clear the log files before you hit the save button in the preferences, then provide the __full__ logfiles with log level __DEBUG__. Replace your domain name / username if you want, but please be sure not to strip crucial information)
  - Please install the carddav plugin using the installation instructions with composer. Especially, delete the `plugins/carddav/vendor` directory if there is one and run composer in the roundcube root directory. Your logs show that you use Guzzle 7, but roundcube 1.5 uses guzzle itself and requires a 6.5 version. If you use composer as described in the install instructions, it will take care of this and you will only have one guzzle library installed. Maybe there is a conflict with two guzzle libraries installed.

Just for completeness, I also tried with radicale and roundcube 1.5beta, it still works as expected. I don't have PHP8 though, but as I wrote I don't expect an issue on that end.
